src/main/java/org/openrewrite/staticanalysis/RemoveUnusedPrivateMethods.java

```diff
@@ -75,53 +75,100 @@ private boolean unusedWarningsSuppressed(J classDeclaration) {
             @Override
             public J.@Nullable MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
                 J.MethodDeclaration m = super.visitMethodDeclaration(method, ctx);
+
                 JavaType.Method methodType = method.getMethodType();
-                if (methodType != null && methodType.hasFlags(Flag.Private) &&
-                        !method.isConstructor() &&
-                        service(AnnotationService.class).getAllAnnotations(getCursor()).isEmpty()) {
+                if (methodType == null) {
+                    return m;
+                }
+
+                // Only consider private, non-constructors.
+                if (!methodType.hasFlags(Flag.Private) || method.isConstructor()) {
+                    return m;
+                }
+
+                // Serialization hooks & similar: do not remove even if private.
+                switch (m.getSimpleName()) {
+                    case "readObject":
+                    case "readObjectNoData":
+                    case "readResolve":
+                    case "writeObject":
+                    case "writeReplace":
+                        return m;
+                }
 
-                    J.ClassDeclaration classDeclaration = getCursor().firstEnclosing(J.ClassDeclaration.class);
-                    if (classDeclaration == null) {
+                // If we're missing the enclosing class or CU, bail out.
+                J.ClassDeclaration classDeclaration = getCursor().firstEnclosing(J.ClassDeclaration.class);
+                if (classDeclaration == null) {
+                    return m;
+                }
+                JavaSourceFile cu = getCursor().firstEnclosingOrThrow(JavaSourceFile.class);
+
+                // If referenced anywhere in TypesInUse (e.g., via other classes), keep it.
+                for (JavaType.Method usedMethodType : cu.getTypesInUse().getUsedMethods()) {
+                    if (methodType.equals(usedMethodType)) {
```

Comment:
Why are you changing unrelated lines here and above? This removal will lead to worse performance, based on
- https://github.com/openrewrite/rewrite/blob/8ce5308a9557dd6ca91253a5c8fb7e6c34a0ef5a/rewrite-java/src/main/java/org/openrewrite/java/tree/JavaType.java#L1563-L1572
- https://github.com/openrewrite/rewrite/blob/8ce5308a9557dd6ca91253a5c8fb7e6c34a0ef5a/rewrite-java/src/main/java/org/openrewrite/java/tree/JavaType.java#L576-L583

Reply:
thanks for the hint, yes will try to undo